loggingservice.go

```diff
@@ -25,36 +26,84 @@ type LoggerInterface interface {
 	Flush() error
 }
 
-type StackdriverClient struct {
+type LoggingClient struct {
+	mtx    sync.Mutex
 	client *logging.Client
 }
 
-func newStackdriverClient(client *logging.Client) LoggerClientInterface {
-	return &StackdriverClient{
-		client: client,
+var globalLoggingClient LoggingClient
+
+func (lc *LoggingClient) checkClientInitialization() error {
+	if lc.client == nil {
+		return fmt.Errorf("logging client is not initialized")
 	}
+	return nil
 }
 
-// Logger creates a new logger to upload log entries
-func (c StackdriverClient) Logger(logID string, opts ...logging.LoggerOption) LoggerInterface {
-	return c.client.Logger(logID, opts...)
+func (lc *LoggingClient) Logger(logID string, opts ...logging.LoggerOption) LoggerInterface {
+	if err := lc.checkClientInitialization(); err != nil {
+		panic(err)
+	}
+	return lc.client.Logger(logID, opts...)
 }
 
-// SetUpOnError sets up stackdriver logging to print errors to the Stderr whenever an internal error occurs on the logger
-func (c StackdriverClient) SetUpOnError() {
-	c.client.OnError = func(e error) {
+func (lc *LoggingClient) SetUpOnError() {
+	if err := lc.checkClientInitialization(); err != nil {
+		panic(err)
+	}
+	lc.client.OnError = func(e error) {
 		fmt.Fprintf(os.Stderr, "logging error: %v", e)
 	}
 }
 
-// Ping will ping the stackdriver log
-func (c StackdriverClient) Ping(ctx context.Context) error {
-	return c.client.Ping(ctx)
+func (lc *LoggingClient) Ping(ctx context.Context) error {
+	if err := lc.checkClientInitialization(); err != nil {
+		return err
+	}
+	return lc.client.Ping(ctx)
+}
+
+func (lc *LoggingClient) Close() error {
+	if err := lc.checkClientInitialization(); err != nil {
+		// Closing the logger here, if it's not initialized we don't care
+		return nil
+	}
+	lc.mtx.Lock()
+	defer lc.mtx.Unlock()
+	err := lc.client.Close()
+	lc.client = nil
+	return err
+}
+
+// CloseGlobalLoggingClient closes the global logging client.
+// This should only be called once, only during a shutdown routine.
+// Calling this while logging is still happening will produce panics
+func CloseGlobalLoggingClient() error {
+	return globalLoggingClient.Close()
```

Comment:
Is there a possibility where we could legitimately have a goroutine still running and logging when our Shutdown function calls `CloseGlobalLoggingClient`? If so, we could panic even after closing (legitmately).

Another question: if we call `CloseGlobalLoggingClient` in shutdown and an errant request comes in that calls `GetOrCreateLoggingClient` then you have a new client that's not flushed.